### PR TITLE
feat(new source): New `generator` source

### PR DESCRIPTION
**Pull request by bruceg**
Ref #1317
Ref #980

Signed-off-by: Bruce Guenter <bruce@timber.io>

This takes some elements of both the above designs, mostly just for simplicity. I am open to alternative designs, I just wanted to make a high speed generator that adds sequence numbers.

I'm not clear what should go into `.meta/links.toml` for the `generator` link. It doesn't fit the pattern of all the other sources which reference some external entity like "standard input" or a data producer.

### CODE REVIEW

File: config/vector.spec.toml
Comment (binarylogic) [inline config/vector.spec.toml:393]: This is not worth blocking the PR, but it would be nice to templatize this field so users should generate formatted test data. For example:

```
{{ timestamp }} - {{ host }} - {{ sequence }}
```

Libraries like [`faker`](https://github.com/cksac/fake-rs) could make this easy.

Comment (bruceg) [reply config/vector.spec.toml:393]: Absolutely, that would be a great way of avoiding needing the sequence number setting.

File: src/sources/generator.rs
Comment (a-rodin) [inline src/sources/generator.rs:21]: One possible extension is to support reading the lines from a file, in addition to the possibility to specify it in the configuration file. This would differ from the `file` source because after the file is read to the end, it would be read from start again, indefinitely.

File: src/sources/generator.rs
Comment (a-rodin) [inline src/sources/generator.rs:23]: When I initially saw the name of the `interval` option, I though that it is the interval between individual events instead of batches. Maybe it could make sense to consider naming it `batch_interval` to indicate that it is the interval between batches and not events, or actually make it an interval between individual events instead of batches?

It also might be useful to have an option to generate events with random intervals between them instead of fixed ones. These random intervals between subsequent events can be sampled using [`rand_distr::Exp`](https://docs.rs/rand_distr/0.2.2/rand_distr/struct.Exp.html) function, which would result in the total time it takes to generate a batch of events being distributed according to the [Erlang distribution](https://en.wikipedia.org/wiki/Erlang_distribution).

File: src/sources/generator.rs
Comment (bruceg) [reply src/sources/generator.rs:21]: That's a great idea. So I'm clear, are you suggesting to keep the file open and keep re-reading from it, or just read it once at initialization and dump those lines?

File: src/sources/generator.rs
Comment (bruceg) [reply src/sources/generator.rs:23]: I am open to changing the name. Issue #1317 calls is `repeat`.

Regarding random intervals, it sounds useful. I'd prefer to do that in a separate PR though.

File: src/sources/generator.rs
Comment (a-rodin) [reply src/sources/generator.rs:21]: I think the first approach is more universal, as it supports arbitrary large files without the risk of going out of memory. In order to make it fast for small files, it might be possible to implement it using `mmap`.

Comment (binarylogic) [reply src/sources/generator.rs:21]: I like the file idea too! But I think we should forgo adding that right now. I assume that would complicate things. Also, the name `generator` implies that it's generating logs, not reading them.

File: config/vector.spec.toml
Comment (binarylogic) [reply config/vector.spec.toml:393]: What do you think then? I think this would be a worthwhile upgrade.

Comment (bruceg) [reply config/vector.spec.toml:393]: Do you want it for the initial implementation or as a follow up PR?

Comment (binarylogic) [reply config/vector.spec.toml:393]: Follow up is fine.

**Comment by bruceg** [comment]
I think this is ready to merge except for the issue with the `generator` entry in `.meta/links.toml`.

**Comment by binarylogic** [comment]
Docs are fixed and look good. I'd like to get an approval from someone on the code though.

**Review by binarylogic — APPROVED** [review]

**Comment by binarylogic** [comment]
@lukesteensen I've assigned you because I'd like to use this in some benchmarks. Could you review the code to see if it's suitable for that? `generator` -> `blackhole` should, in theory, be a very fast pipeline to get baseline performance numbers.

**Comment by bruceg** [comment]
I believe the test failures are just out of disk space errors.

**Comment by binarylogic** [comment]
I'm rerunning the jobs to see if it's intermittent. I hope it isn't related to this change.

**Review by lukesteensen — APPROVED** [review]

**Comment by lukesteensen** [comment]
We should be able to benchmark this on its own to see what kind of throughput it will output. I don't see anything obvious that should make it unsuitable for testing other components.

**Pull request merged**
